### ISSUE TITLE
Performance Optimizations

### ISSUE THREAD

**Issue by pitbox46** [gisellevonbingen-Minecraft/Ad-Astra-Giselle-Addon#57]
### The details

Forge 1.20.1, ATM9

This mod adds quite a bit of overhead without even using it's features. The biggest issue is checking and attempting to distribute oxygen (`OxygenChargerUtils#distributeToItems()`) for every entity of every tick. I will be changing the logic so it runs only once every 20 ticks and distributes 20 times as much (so I can quickly deploy a fix for a server).  
However, this is only a bandage since it looks like in `streamExtractable()` that this implementation would leave some oxygen left over in the tanks. I'm trying to think through it, but I think it would require some code reorganization to make it work nicely.

Also `ProofAbstractUtils` seems to be using a non-insignificant amount of time as well. This seems to be due to the use of CompoundTags. It looks like tags aren't a very efficient way to store commonly accessed data.

![image](https://github.com/gisellevonbingen-Minecraft/Ad-Astra-Giselle-Addon/assets/37819783/0fb16cf9-02b1-4bf6-a71f-6af6be896f9b)

**Comment by pitbox46** [comment]
If you would like I have the commit I made here. It's very basic, but it cuts the tick times down to 5% of what it was

https://github.com/gisellevonbingen-Minecraft/Ad-Astra-Giselle-Addon/compare/1.20.1-rework...pitbox46:Ad-Astra-Giselle-Addon:1.20.1-rework

**Comment by gisellevonbingen** [comment]
Looks good, i will patch thier soon.

**Comment by gisellevonbingen** [comment]
ok, i added checking delay condition for oxygen distribution.
to modify config, can reduce cpu overhead more.

and i agree what you said nbt is not efficient.
there too many allocations and calls to just save integers.
so, i fixed to used nbt at load/save entity data only.

**Comment by item4** [comment]
@gisellevonbingen Hi! This may be worth a quick recheck on the 1.21.1 port.

While profiling version 8.1, I noticed that `OxygenChargerUtils.distributeToItems()` can still become a noticeable server-thread path depending on the items carried by the player.

In this case, `OxygenStorageUtils.isInfinifySource()` was querying `FluidApi.ITEM` for inventory items, and a Sophisticated Backpack with the Inception Upgrade made the capability lookup more expensive:

```
OxygenChargerUtils.distributeToItems()
└─ OxygenStorageUtils.isInfinifySource()
   └─ FluidItemLookup.find()
      └─ ItemCapability.getCapability()
         └─ BackpackWrapper.getItemFluidHandler()
            └─ BackpackWrapper.getFluidHandler()
               └─ UpgradeHandler.getWrappersThatImplement()
```

In two consecutive five-minute profiles, this path accounted for approximately 1.52% and 1.47% of the server thread. After moving the backpack out of the player's inventory, it dropped to approximately 0.06%.

I also reproduced the same interaction on an unchanged ATM10 8.1 instance, although the absolute cost there was small.

I'm not sure whether any change is needed here or in Sophisticated Backpacks, since the addon appears to be making a normal fluid-capability query. I thought it might still be useful to mention here in case the polling behavior or interval is worth reviewing for the 1.21.1 port.

Related ATM10 report: AllTheMods/ATM-10#4364

**Comment by gisellevonbingen** [comment]
Since this issue has already been closed and the cause is different, it would be better to create a new issue next time. :)

https://github.com/gisellevonbingen-Minecraft/Ad-Astra-Giselle-Addon/blob/aa9c1d93739e768b401515890626aa3c896c784e/common/src/main/java/ad_astra_giselle_addon/common/content/oxygen/OxygenStorageUtils.java#L126-L129

In fact, in this code, `CreativeFluidContainer` is used only in this mod.
Separate from `Sophisticated Backpacks`, it would be good to simplify the expression in this mod as well.

**Comment by item4** [comment]
Thank you for checking and updating it so quickly!
Understood — I’ll open a separate issue next time when the cause differs from an older closed issue. The more specific item check makes sense, and I’ll retest it once a build containing the change is available.
